### ISSUE TITLE
Transaction update RowsAffected is 0

### ISSUE THREAD

**Issue by aimuz** [go-gorm/gorm#2038]
Transaction update RowsAffected is 0

Just like the following code, when it updates, RowsAffected is 0

### What version of Go are you using (`go version`)?

`go version go1.10.3 darwin/amd64`

### Which database and its version are you using?

mysql 5.7

```go
package main

import (
	"github.com/jinzhu/gorm"
	_ "github.com/jinzhu/gorm/dialects/mysql"
	"log"
)

var db *gorm.DB

func init() {
	var err error
	db, err = gorm.Open("mysql", "gorm:gorm@tcp(localhost:9910)/gorm?charset=utf8&parseTime=True")
	if err != nil {
		panic(err)
	}
	db.LogMode(true)
}

type UserTest struct {
	gorm.Model

	Name     string
	Password string
}

func main() {
	ret := db.AutoMigrate(&UserTest{})
	if ret.Error != nil {
		log.Fatal(ret.Error)
	}

	user := &UserTest{
		Name:     "5",
		Password: "2",
	}
	ret = db.Model(&UserTest{}).Create(user)
	if ret.Error != nil {
		log.Fatal(ret.Error)
	}

	var users []UserTest
	ret = db.Model(&UserTest{}).Find(&users)
	if ret.Error != nil {
		log.Fatal(ret.Error)
		return
	}

	for _, v := range users {
		log.Println(v.Name, v.Password)
	}

	tx := db.Begin()

	ret = tx.Table("user_tests").Where("name = ?", "5").Updates(map[string]interface{}{
		"password": "12345",
	})
	if ret.Error != nil {
		log.Fatal(ret.Error)
		tx.Rollback()
		return
	}
	log.Println(ret.RowsAffected)

	ret = tx.Table("user_tests").Where("name = ?", "4").Updates(map[string]interface{}{
		"password": "44444",
	})
	if ret.Error != nil {
		log.Fatal(ret.Error)
		tx.Rollback()
		return
	}
	log.Println(ret.RowsAffected)
	tx.Commit()

	// 重新查
	ret = db.Model(&UserTest{}).Find(&users)
	if ret.Error != nil {
		log.Fatal(ret.Error)
		return
	}
	for _, v := range users {
		log.Println(v.Name, v.Password)
	}
}
```

<img width="1602" alt="wx20180815-105911 2x" src="https://user-images.githubusercontent.com/17666471/44129352-5a6c0e44-a07a-11e8-978f-f1feaa4568d3.png">

**Comment by pedromorgan** [comment]
I can't see what the problem (but am stupid), can u provide more detail if possible..

Is this a bug?

**Comment by aimuz** [comment]
Yes, I think it's a Bug, and this bug's problem leads to the fact that when you update a record the same as a database, you don't get a return record.

**Comment by aimuz** [comment]
The most important thing is that this Bug's lead occurs only when the transaction occurs, and when the update command is executed separately, the number of rows that can be updated is obtained. I'm puzzled

**Comment by alicksnake22** [comment]
@aimuz no data with name = 4

**Comment by Alllfred** [comment]
Met the same problem

**Comment by h2ero** [comment]
same problem. update is success but RowsAffected is 0

**Comment by aimuz** [comment]
I don't know if this project is stopped.

**Comment by kutysam** [comment]
This is affecting me too. Even for delete within a transaction.

**Comment by caojunxyz** [comment]
I got the same problem

**Comment by AlessandroSechi** [comment]
Same problem also for me on
`go version go1.13.1 linux/amd64`
and
`github.com/jinzhu/gorm v1.9.11`

**Comment by bluebrown** [comment]
https://i.imgur.com/8sZpqld.png

**Comment by bluebrown** [comment]
i have the same. db log says one row affected  but db.AffectedRows is zero.

**Comment by HsiaohuiHsiang** [comment]
serious bug

**Comment by seon1-kim** [comment]
this issue isn't solved?

**Comment by aimuz** [comment]
Yes, it hasn't been solved yet

**Comment by wilian746** [comment]
This is happening to me also in the query with the database in memory

`SELECT * FROM "expressions"  WHERE ("expressions"."id" = 'ff0f834b-370a-4836-a612-6ec72a8bfd37') ORDER BY "expressions"."id" ASC LIMIT 1  
[1 rows affected or returned ] `

`result.RowsAffected` <- the value is ZERO

**Comment by doscode-kr** [comment]
I got same problem, too

**Comment by CHURLZ** [comment]
For others coming to this thread, I could verify this works for sqlite3 and postgres.

	if ret.RowsAffected == 0 {
		tx.Rollback()
		return fmt.Errorf("no record found for id %s and status CREATED", _feedback.Id)
	} else if ret.RowsAffected > 1 {
		tx.Rollback()
		return fmt.Errorf("update affected %d rows, maximum allowed = 1", ret.RowsAffected)
	}

**Comment by eruca** [comment]
me too

**Comment by yiippee** [comment]
have plan to fix this bug? or there is aother way to know the RowsAffected. Thanks a lot.

**Comment by ovdmar** [comment]
It happened to me too.

**Comment by Bulesxz** [comment]
![image](https://user-images.githubusercontent.com/11548849/83776450-c526be80-a6ba-11ea-9e84-f0aa113a2940.png)

Your question does not match your description

**Comment by Bulesxz** [comment]
> I got the same problem. I find the affected in db debug log is right
> 
> ```
> UPDATE `orders` SET `status` = 2, `updated_at` = '2019-07-11 19:08:08'  WHERE (order_no = 155765990288785421 and status = 1)  
> [1 rows affected or returned ] 
> ```
> 
> But zero (not right) in return:
> 
> ```
> tx.Model(&tradeModel.Order{}).
> 		Where("order_no = ? and status = ?", orderNo, int16(pb.OrderStatus_OrderStatusWaitConfirm)).
> 		Update("status", int16(status))
> log.Println(tx.RowsAffected) // is zero 
> ```


tx.RowsAffected  Incorrect usage 

ret = tx.Model(&tradeModel.Order{}).
		Where("order_no = ? and status = ?", orderNo, int16(pb.OrderStatus_OrderStatusWaitConfirm)).
		Update("status", int16(status))
log.Println(ret.RowsAffected) 

ret.RowsAffected  != tx.RowsAffected

**Comment by Bulesxz** [comment]
tx.RowsAffected  Incorrect usage
tx not RowsAffected

**Comment by aimuz** [comment]
> tx.RowsAffected Incorrect usage
> tx not RowsAffected

So how to use it

It seems that I'm not alone in this problem

**Comment by Bulesxz** [comment]
> > tx.RowsAffected Incorrect usage
> > tx not RowsAffected
> 
> So how to use it
> 
> It seems that I'm not alone in this problem

tx.RowsAffected Incorrect usage

ret = tx.Model(&tradeModel.Order{}).
Where("order_no = ? and status = ?", orderNo, int16(pb.OrderStatus_OrderStatusWaitConfirm)).
Update("status", int16(status))
log.Println(ret.RowsAffected)

ret.RowsAffected != tx.RowsAffected


ret replace tx

**Comment by aimuz** [comment]
> > > tx.RowsAffected Incorrect usage
> > > tx not RowsAffected
> > 
> > 
> > So how to use it
> > It seems that I'm not alone in this problem
> 
> tx.RowsAffected Incorrect usage
> 
> ret = tx.Model(&tradeModel.Order{}).
> Where("order_no = ? and status = ?", orderNo, int16(pb.OrderStatus_OrderStatusWaitConfirm)).
> Update("status", int16(status))
> log.Println(ret.RowsAffected)
> 
> ret.RowsAffected != tx.RowsAffected
> 
> ret replace tx

I don't think you understand me

Although I haven't used it for a long time, by checking my history

There are two updates in the transaction. The first update can get the rowsaffected, while the second update cannot get the correct rowsaffected

If I were to ask questions at that time, I might have more details.

**Comment by Bulesxz** [comment]
> > > > tx.RowsAffected Incorrect usage
> > > > tx not RowsAffected
> > > 
> > > 
> > > So how to use it
> > > It seems that I'm not alone in this problem
> > 
> > 
> > tx.RowsAffected Incorrect usage
> > ret = tx.Model(&tradeModel.Order{}).
> > Where("order_no = ? and status = ?", orderNo, int16(pb.OrderStatus_OrderStatusWaitConfirm)).
> > Update("status", int16(status))
> > log.Println(ret.RowsAffected)
> > ret.RowsAffected != tx.RowsAffected
> > ret replace tx
> 
> I don't think you understand me
> 
> Although I haven't used it for a long time, by checking my history
> 
> There are two updates in the transaction. The first update can get the rowsaffected, while the second update cannot get the correct rowsaffected
> 
> If I were to ask questions at that time, I might have more details.

![image](https://user-images.githubusercontent.com/11548849/84997828-9f50ed80-b181-11ea-9edd-8354437e473b.png)

从你的截图里面看，是你使用的方法不对
不能使用  tx.RowsAffected 这个是个全局的，应该使用ret.RowsAffected (ret 是每一次操作单独调用结果)

**Comment by hktalent** [comment]
me too
xx1 := GetSession().Exec(szSql, args...)
xx1.RowsAffected always is 0

**Comment by zhoulin1359** [comment]
`{"cost":0.003117164,"file":"/www/host/blockchain-price/src/infra/repo/currency_repo.go:77","level":"info","msg":"","request_id":"iZwz9heg0pec2ut18ixomcZ.1663682053.1572222602964398080","rows":0,"sql":"UPDATE `currency` SET `day_change`=-4.930000,`day_change_amount`=-0.084000,`high_price`=1.920000,`low_price`=1.583000,`price`=1.620000,`update_time`='2022-09-20 21:54:14.409' WHERE `symbol` = 'NEBL'","time":"2022-09-20T21:54:14+08:00"}
{"cost":0.003309807,"file":"/www/host/blockchain-price/src/infra/repo/currency_repo.go:77","level":"info","msg":"","request_id":"iZwz9heg0pec2ut18ixomcZ.1663682052.1572222599189524480","rows":1,"sql":"UPDATE `currency` SET `day_change`=-4.930000,`day_change_amount`=-0.084000,`high_price`=1.920000,`low_price`=1.583000,`price`=1.620000,`update_time`='2022-09-20 21:54:13.595' WHERE `symbol` = 'NEBL'","time":"2022-09-20T21:54:13+08:00"}
`

这是我日志记录的sql，第一个记得rows为0，第二个记录rows为1，这区间数据库没有任何改动，加上 update_time 是为了保证数据库有值就一定变更，结果是数据库变更了，RowsAffected还是为0

**Comment by akshay-pj-open** [comment]
I'm also facing this issue, even though it actually inserts the records. Kindly note that for me this bug arises only during [batch inserts](https://gorm.io/docs/create.html#Batch-Insert).

**Comment by LemonNekoGH** [comment]
> > > > > tx.RowsAffected Incorrect usage
> > > > > tx not RowsAffected
> > > > 
> > > > 
> > > > So how to use it
> > > > It seems that I'm not alone in this problem
> > > 
> > > 
> > > tx.RowsAffected Incorrect usage
> > > ret = tx.Model(&tradeModel.Order{}).
> > > Where("order_no = ? and status = ?", orderNo, int16(pb.OrderStatus_OrderStatusWaitConfirm)).
> > > Update("status", int16(status))
> > > log.Println(ret.RowsAffected)
> > > ret.RowsAffected != tx.RowsAffected
> > > ret replace tx
> > 
> > 
> > I don't think you understand me
> > Although I haven't used it for a long time, by checking my history
> > There are two updates in the transaction. The first update can get the rowsaffected, while the second update cannot get the correct rowsaffected
> > If I were to ask questions at that time, I might have more details.
> 
> ![image](https://user-images.githubusercontent.com/11548849/84997828-9f50ed80-b181-11ea-9edd-8354437e473b.png)
> 
> 从你的截图里面看，是你使用的方法不对 不能使用 tx.RowsAffected 这个是个全局的，应该使用ret.RowsAffected (ret 是每一次操作单独调用结果)

谢谢，解决了
Issue solved, thank you

**Comment by Fantomo** [comment]
It’s 2024 and this bug hasn’t been solved yet

**Comment by rea1shane** [comment]
When I run `info, err := d.Where(d.ID.Eq(1)).Update(d.Status, 1)`, sometimes `info.RowsAffected` is `1`, sometimes is `0`. But even though the value is `0`, the database is updated successfully.

**Comment by strconv** [comment]
```go
//  wrong case
tx := xx.Begin()
tx.Update(xx).Error
fmt.Println(tx.RowsAffected)   // result is 0

// correct case
tx := xx.Begin()
db := tx.Update(xx)
fmt.Println(db.RowsAffected)   // result is n
```

**Comment by BLN-techin-phi** [comment]
this might not related that much but please allow me to share
about an hour ago, I'm also faced `0 rowsAffected` on update some of my data


### TLDR; if data is not existed (or was soft-deleted), <br> trying to update them also make you get `0 rowsAffected`

after go through each comment of this issue
I ensured that my code call `RowsAffected`
from the result of `tx.Update(...)` not `tx.RowsAffected`
(using result, from `result = tx.Update(...)` - like the above comment)
but error still the same right there

finally, after debug more a little bit,
I found that update function return the correct result
but what cause error is the data in the database

here is pseudo code that might help to understand what happen
```
   // these all below are inside transaction
  _ = deleteSomethingByIds(ctx, ids)
  _ = updateSomething(ctx, pk, mapDataToUpdate)
```

I did send empty list of id (the `ids`) to function to delete items by ids
so Gorm soft-delete every records which make my next query,
that try to update -> found nothing (due to it respect soft-delete boolean flag)
